jscomp/bsb/bsb_helper_main.ml

```diff
@@ -34,7 +68,154 @@ let () =
+ * out the dependencies graph to do a topological sort before calling ocamlc. 
+ * 
+ * Below we check link_flag first, then check pack_flag.
+ * *)
+let () = match !link_flag with
+| Some link_byte_or_native ->
+  begin match !main_module with
+  | None -> failwith "linking needs a main module"
+  | Some main_module ->
+    let suffix_object_files, suffix_library_files, compiler, output_file = begin match link_byte_or_native with
+    | LinkBytecode output_file -> Literals.suffix_cmo, Literals.suffix_cma , "ocamlc.opt"  , output_file
+    | LinkNative output_file   -> Literals.suffix_cmx, Literals.suffix_cmxa, "ocamlopt.opt", output_file
+    end in
+    (* Map used to track the path to the files as the dependency_graph that we're going to read from the mlast file only contains module names *)
+    let module_to_filepath = List.fold_left 
+      (fun m v -> 
+        String_map.add 
+        (Ext_filename.module_name_of_file_if_any v) 
+        (Ext_filename.chop_extension_if_any v) 
+        m) 
+      String_map.empty 
+      !batch_files in
+    let dependency_graph = List.fold_left
+      (fun m file -> 
+        String_map.add 
+          (Ext_filename.module_name_of_file_if_any file)
+          (Bsb_helper_extract.read_dependency_graph_from_mlast_file ((Ext_filename.chop_extension file) ^ Literals.suffix_mlast))
+          m)
+      String_map.empty
+      !batch_files in
+    let tasks = Bsb_helper_extract.simple_collect_from_main dependency_graph main_module in
+    let domain = Queue.fold 
+      (fun domain base -> String_set.add base domain)
+      String_set.empty
+      tasks in
+    let sorted_tasks = Bsb_helper_extract.sort_files_by_dependencies ~domain dependency_graph in
+    let list_of_object_files = Queue.fold 
+      (fun acc v -> match String_map.find_opt v module_to_filepath with
+        | Some file -> (file ^ suffix_object_files) :: acc
+        | None -> failwith @@ "build.ninja is missing the file '" ^ v ^ "' that was used in the project. Try force-regenerating but this shouldn't happen."
+        )
+      []
+      sorted_tasks in
+    (* This list will be reversed so we append the otherlibs object files at the end, and they'll end at the beginning. *)
+    let all_object_files = List.rev (list_of_object_files @ (get_otherlibs_dependencies dependency_graph suffix_library_files)) in
+    if List.length all_object_files > 0 then begin
+      let library_files = List.fold_left
+        (fun acc dir ->
+          (Ext_filename.combine dir (Literals.library_file ^ suffix_library_files)) :: acc)
+        [] !includes in
+      Unix.execvp
+        compiler
+        (Array.of_list (compiler :: "-o" :: output_file :: library_files @ all_object_files))
+    end else
+      failwith @@ "No " ^ suffix_object_files ^ " to link. Hint: is the main module in the entries array right?"
+  end
+| None -> ()
 
-    ] annoymous usage
+let () = match !pack_flag with
+| Some pack_byte_or_native ->
```

Comment:
This isn't for -pack, sorry if this is misleading.
This is for packing an external dependency into a `.a` or `.cmxa` file. Does it make sense?